vsphere/resource_vsphere_vmfs_datastore.go

```diff
@@ -0,0 +1,324 @@
+package vsphere
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+const (
+	retryDeletePending   = "retryDeletePending"
+	retryDeleteCompleted = "retryDeleteCompleted"
+	retryDeleteError     = "retryDeleteError"
+
+	waitForDeletePending   = "waitForDeletePending"
+	waitForDeleteCompleted = "waitForDeleteCompleted"
+	waitForDeleteError     = "waitForDeleteError"
+)
+
+// formatCreateRollbackError defines the verbose error for extending a disk on
+// creation where rollback is not possible.
+const formatCreateRollbackErrorUpdate = `
+WARNING: Dangling resource!
+There was an error extending your datastore with disk: %s:
+%s
+Additionally, there was an error removing the created datastore:
+%s
+You will need to remove this datastore manually before trying again.
+`
+
+// formatCreateRollbackError defines the verbose error for extending a disk on
+// creation where rollback is not possible.
+const formatCreateRollbackErrorProperties = `
+WARNING: Dangling resource!
+After creating the datastore, there was an error fetching its properties:
+%s
+Additionally, there was an error removing the created datastore:
+%s
+You will need to remove this datastore manually before trying again.
+`
+
+// formatUpdateInconsistentState defines the verbose error when the setting
+// state failed in the middle of an update opeartion. This is an error that
+// will require repair of the state before TF can continue.
+const formatUpdateInconsistentState = `
```

Comment:
could we not make use of `d.Partial()` to continue & retry any pending steps?

Reply:
This isn't necessarily a problem with state that was already set, but more with state that could _not_ be set. When we are adding extents, we need to add the extent, then update the state in real time so that in the event that some of the disks were added, but some were not, the disks that were already added are properly saved to state, and we can pick up where we left off on the next Terraform run. This "inconsistent state" error is to catch the edge case where, for some reason, we could not set the `TypeList` properly and hence we are left with a situation were the disk was added, but the state was not updated, at which point the state would need to be manually repaired (although a `terraform refresh`) could actually fix it - I've updated the error message to reflect this.